### DISCUSSION TITLE
Livewire Message 404

### DISCUSSION THREAD

**Discussion by striker4150** [livewire/livewire#5557] [1. Help]
Hi guys,

This requires a bit of background explanation. I'm running Laravel 10 with Vite and Livewire. It's being run in a subdirectory, behind a reverse proxy (e.g. `http://www.mydomain.com/subdir/project` -> `http://192.168.x.x:8083/subdir/project/public`). Here are the configurations I made to do so:

Apache
```
ProxyPass /subdir/project http://192.168.x.x:8083/subdir/project/public
ProxyPassReverse /subdir/project http://192.168.x.x:8083/subdir/project/public
```

.env
```
APP_URL=http://www.mydomain.com/subdir/project
ASSET_URL=/subdir/project
LIVEWIRE_ASSET_URL=/subdir/project
```

vite.config.js
```js
import { defineConfig } from 'vite';
import laravel from 'laravel-vite-plugin';

export default defineConfig({
    base: '/subdir/project/build/',
    plugins: [
        laravel({
            input: ['resources/css/app.css', 'resources/js/app.js'],
            refresh: true,
        }),
    ],
});
```

AppServiceProvider.php
```php
/**
 * Bootstrap any application services.
 */
public function boot(): void {
    URL::forceRootUrl(config('app.url'));
}
```

I'm trying to use the Laravel-Table library's "Table" livewire component, but the livewire POST request to `http://www.mydomain.com/subdir/project/livewire/message/table` is 404'ing. Any ideas on what might be causing this? Everything else seems to be routed correctly, including `http://www.mydomain.com/subdir/project/livewire/livewire.js`.

![Network Requests](https://user-images.githubusercontent.com/8561358/219224112-f2d97e24-4f9b-483f-8c38-08521827db52.png)

![Website](https://user-images.githubusercontent.com/8561358/219224120-864fddf8-64fc-4cb6-8184-fdbdc38623e4.png)

**Comment by joshhanley** [comment]
@striker4150 please don't open both an issue and a discussion 🙂

In regards to your issue, what patch version of Livewire are you using? Run `composer show -D` in your console and look for Livewire.

**Reply by striker4150** [reply]
Ah, my bad. I'm using v2.11.2.

**Reply by joshhanley** [reply]
@striker4150 no worries! Thanks, hmm so we recently patched some issues that were causing the 404. But I suspect that it may be triggering for your use case due to the proxy.

But before we go down that route, can you provide some more details about how it's structured? What reverse proxy are you using (and why)? Also, are you passing the Livewire headers through the proxy? Is this a local set up or on servers/ in production? If you can add any other detail you can think of that may help too, would be good.

**Reply by striker4150** [reply]
The project is hosted on a remote dev server (not local). Requests to the domain go through an Apache reverse proxy, where they are ProxyPassed to the Apache server that actually hosts the project files. The ProxyPass directives that are actually used can be found in in my original post. The proxied requests go to a VirtualHost on *:8083, and the associated scripts are executed through FCGI. I haven't heard about any Livewire headers though. Could that possibly be the issue?

**Reply by joshhanley** [reply]
@striker4150 ok thanks, yeah sounds like it might be the header. There is a `X-Livewire` header that will need to be passed through the proxy to the app server. I'm pretty sure there is no value, it just needs to exist. Check a network request out locally and you should see it in the request headers.

Add that in and I think it should solve the issue 🙂

**Reply by striker4150** [reply]
Thanks, I'll test it out first thing tomorrow.

**Reply by joshhanley** [reply]
@striker4150 no worries! Let me know how it goes.

**Reply by striker4150** [reply]
@joshhanley Hey, I just tested it, and it seems like the `X-Livewire` header is successfully passing through the proxy. I checked it by making a request to the route:
```php
Route::get('/test', function(\Illuminate\Http\Request $request) {
    return $request->headers->all();
});
```
so I'm pretty confident that it went through. It's still returning 404.

**Reply by joshhanley** [reply]
@striker4150 I don't follow how that route test would help, the Livewire header should only be available on Livewire routes (`/livewire/message').

But if you think it's going through, then what you can do is modify the below Livewire file and see if you can log what path it is trying to find. 

https://github.com/livewire/livewire/blob/e938cfc9da88d49d1e27eedcd79879f7fdeb2371/src/Controllers/HttpConnectionHandler.php#L82-L89

Above line 84 in the file above, log out the `$url` and `$method` so we can see what route path it's trying to match too, then you can compare that with your routes list and see what is wrong.

**Reply by striker4150** [reply]
I used the test route by sending a modified request with the `X-Livewire` header attached. If the header went through the proxy correctly, it would show up in the response, and it did.

As for the logging, the headers are going through, as I can see the `[HTTP_X_LIVEWIRE] => true` in the `request()->server` object, but it's being constructed as a `GET` request to `http://www.mydomain.com/subdir/project`.

**Reply by joshhanley** [reply]
@striker4150 ah right, yep makes sense! Ok good to know the header is going through.

Sorry, just to clarify, you logged out both those variables above? As they won't be the same as what is in the request.

Also, can you confirm what URL you initially load the component on?

**Reply by joshhanley** [reply]
@striker4150 do you have discord? If so, do you want to DM me `@joshhanley#6022` so we can keep investigating? I will need a dump of the routes generated from a few lines further down too but might be better in private.

**Reply by striker4150** [reply]
Sure, I can do that. I'll message you from striker4150#1484.

**Reply by marispro** [reply]
Hello guys, have you solved there anything? In some component, after livewire/laravel update, I am also getting 404, however other components working good so far.

**Reply by joshhanley** [reply]
@marispro the issue was related to how the project was being hosted in the subdirectory. What was happening was the path was ending up with the `public` directory in it when it shouldn't have, due to how apache was configured.

If you can provide some details about your setup maybe we can try and troubleshoot it?

**Reply by tomphilpotts** [reply]
@joshhanley 

I am running into a similar problem, using a reverse proxy with nginx. I can't find anything about how to set the header correctly. 

Is this the correct way?

```
 location /directory/ {
        proxy_pass http://IP_ADDRESS:80/;
        proxy_http_version 1.1;
        proxy_buffering off;
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;
        proxy_set_header X-Forwarded-Path /directory;
        proxy_set_header X-Forwarded-Prefix "/directory";
        proxy_pass_header X_LIVEWIRE;
        proxy_pass_header X-CSRF-TOKEN;
    }
```
